idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/frontend/api/fir/components/KtFirSmartcastProvider.kt

```diff
@@ -50,4 +54,21 @@ internal class KtFirSmartcastProvider(
             }?.let(::add)
         }
     }
+
+    override fun getSmartCastNullability(expression: KtExpression): KtSmartCastNullability = withValidityAssertion {
```

Comment:
I find it arbitrary to bundle this nullability concept with smartcast. How about moving it into `KtExpressionTypeProvider`? Then the second and third branch can simply be merged together so that this function can also return a sensible value for non-smartcast expressions.

Reply:
But, having separate 2nd/3rd branches allowed us to _not_ call `isNullabltType()` check for non-smartcast expressions?

Reply:
I think what @tgeng means is that for regular `FirExpression`, we can also call `!fir.typeRef.coneType.isNullableType()` to get the "definite nullability", e.g.:
```kt
       return when (val fir = expression.getOrBuildFir(analysisSession.firResolveState)) {
            is FirExpressionWithSmartcastToNull -> KtDefiniteNullability.NULL
            is FirExpression -> {  // Covers smartcast and non-smartcast
                with(analysisSession.rootModuleSession.typeContext) {
if (!fir.typeRef.coneType.isNullableType()) {
                        KtDefiniteNullability.NOT_NULL
                    } else {
                        KtDefiniteNullability.UNKNOWN
                    }
                }
            }
            else -> KtDefiniteNullability.UNKNOWN
        }
```
@tgeng Is this what you meant?

Reply:
Yep this is what I meant.